File: java17/src/main/java/org/chelck/dp/Timings.java
package org.chelck.dp;

import java.util.concurrent.TimeUnit;

public class Timings {
    private long start;
    private long total;
    private long max;

    public Timings() {
        this.start = 0;
        this.total = 0;
        this.max = 0;
    }

    public void start() {
        this.start = System.nanoTime();
    }

    public void mark() {
        long now = System.nanoTime();
        long elapsed = now - this.start;
        this.total += elapsed;
        this.max = Math.max(this.max, elapsed);
        this.start = now;
    }

    public String toString() {
        return "total: " + f(this.total) + ", max: " + f(this.max);
    }

    private long f(long n) {
        return TimeUnit.NANOSECONDS.toMillis(n);
    }
}

File: scala3_mutex/src/main/scala/org/chelck/dp/Timings.scala
package org.chelck.dp

import java.util.concurrent.TimeUnit

case class Timings(var startTime: Long, var total: Long, var max: Long) {

  def start(): Unit = startTime = System.nanoTime();

  def mark(): Unit = {
    val now = System.nanoTime
    val elapsed = now - this.startTime
    this.total = total + elapsed
    this.max = Math.max(this.max, elapsed)
    this.startTime = now
  }
  
  private def f(n: Long) = TimeUnit.NANOSECONDS.toMillis(n)

  override def toString: String = s"total=${f(total)}, max=${f(max)}"

}
